src/ruby/ext/grpc/rb_xds_server_credentials.c

```diff
@@ -0,0 +1,162 @@
+   a peer ruby object, 'mark' to hold references to objects involved in
+   constructing the server credentials. */
+typedef struct grpc_rb_xds_server_credentials {
+  /* Holder of ruby objects involved in constructing the server credentials */
+  VALUE mark;
+  /* The actual server credentials */
+  grpc_server_credentials* wrapped;
+} grpc_rb_xds_server_credentials;
+
+/* Destroys the server credentials instances. */
+static void grpc_rb_xds_server_credentials_free(void* p) {
+  grpc_rb_xds_server_credentials* wrapper = NULL;
+  if (p == NULL) {
+    return;
+  };
+  wrapper = (grpc_rb_xds_server_credentials*)p;
+
+  /* Delete the wrapped object if the mark object is Qnil, which indicates that
+     no other object is the actual owner. */
+  if (wrapper->wrapped != NULL && wrapper->mark == Qnil) {
+    grpc_server_credentials_release(wrapper->wrapped);
+    wrapper->wrapped = NULL;
+  }
+
+  xfree(p);
+}
+
+/* Protects the mark object from GC */
+static void grpc_rb_xds_server_credentials_mark(void* p) {
+  grpc_rb_xds_server_credentials* wrapper = NULL;
+  if (p == NULL) {
+    return;
+  }
+  wrapper = (grpc_rb_xds_server_credentials*)p;
+
+  /* If it's not already cleaned up, mark the mark object */
+  if (wrapper->mark != Qnil) {
+    rb_gc_mark(wrapper->mark);
+  }
+}
+
+static const rb_data_type_t grpc_rb_xds_server_credentials_data_type = {
+    "grpc_xds_server_credentials",
+    {grpc_rb_xds_server_credentials_mark, grpc_rb_xds_server_credentials_free,
+     GRPC_RB_MEMSIZE_UNAVAILABLE, NULL},
+    NULL,
+    NULL,
+#ifdef RUBY_TYPED_FREE_IMMEDIATELY
+    RUBY_TYPED_FREE_IMMEDIATELY
+#endif
+};
+
+/* Allocates ServerCredential instances.
+
+   Provides safe initial defaults for the instance fields. */
+static VALUE grpc_rb_xds_server_credentials_alloc(VALUE cls) {
+  grpc_rb_xds_server_credentials* wrapper =
+      ALLOC(grpc_rb_xds_server_credentials);
+  wrapper->wrapped = NULL;
+  wrapper->mark = Qnil;
+  return TypedData_Wrap_Struct(cls, &grpc_rb_xds_server_credentials_data_type,
+                               wrapper);
+}
+
+/* The attribute used on the mark object to preserve the fallback_creds. */
+static ID id_fallback_creds;
+
+/*
+  call-seq:
+    creds = ServerCredentials.new(fallback_creds)
+
+    fallback_creds: (ServerCredentials) fallback credentials to create
+                    XDS credentials.
+
+    Initializes ServerCredential instances. */
+static VALUE grpc_rb_xds_server_credentials_init(VALUE self,
+                                                 VALUE fallback_creds) {
+  grpc_rb_xds_server_credentials* wrapper = NULL;
+  grpc_server_credentials* creds = NULL;
+
+  grpc_server_credentials* grpc_fallback_creds =
+      grpc_rb_get_wrapped_server_credentials(fallback_creds);
+  creds = grpc_xds_server_credentials_create(grpc_fallback_creds);
+
+  if (creds == NULL) {
+    rb_raise(rb_eRuntimeError, "could not create a credentials, not sure why");
```

Comment:
similar comment about specifying the error message as for the `rb_xds_channel_credentials.c` file

Reply:
Same as previous